### PR TITLE
Lots of improvements

### PR DESCRIPTION
**Pull request by BillyDM**
I was looking over this code last week and I noticed a lot of things I wanted to improve. Here is the list of changes I made:

* Removed all instances of `unsafe`.
* Fixed an issue where the write stream may not write the last remaining samples in the buffer when closed.
* Made the server polling rate configurable.
* Used a single `AudioBlock` type for both the read and write streams.
* Improved the interleaving performance in the WAV writer and the demos.
* Removed the unnecessary interleaving/deinterleaving steps in the Symphonia decoder. Symphonia already has its buffers in deinterleaved form, so removing this should improve performance.
* Renamed a lot of things, and made some non-relevant things private.
* Edited the readme.
* Added and edited some documentation.
* Renamed the `new` method to `spawn` on the servers.
* Updated the demos to use the latest version of egui (and a small color change on the player demo for readability).
* Made the write demo export to the ./target folder.
* Removed the declicking code in the demo player because it didn't work anyway.
* Used more Rust-isms like iterators in some places.
* Bumped versions on Cargo.toml.
* Renamed the `decode` feature to `decode-symphonia`.
* Exclude the test audio files from being included in the crates.io package.

Needless to say, this contains a lot of breaking changes.

### CODE REVIEW

**Comment by uklotzde** [comment]
Please change the PR merge strategy of the repo to "Allow squash merging" only to avoid merging all those meaningless, intermediate commits:

image

This will create a linear and streamlined commit history.

**Comment by uklotzde** [comment]
Too many changes for a proper review. I suggest to simply merge this PR and fix any open issues afterwards.

Tagging this repo with "1.0" was too early anyway.

**Comment by BillyDM** [comment]
Okay, I'll allow squashing commits.

Also yeah, I probably jumped the gun a bit by tagging it "1.0". Would it be possible to revert that?

**Comment by BillyDM** [comment]
Hmm, there doesn't seem to be any way to revert to a previous version on crates.io. Oh well, I'll just go ahead and merge it.

**Comment by BillyDM** [comment]
Actually, I'll hear from @Be-ing first.

File: core/src/read/mod.rs [hunk 1]
```diff
@@ -71,15 +76,17 @@ pub struct ReadStreamOptions<D: Decoder> {
...
     /// This should be left alone unless you know what you are doing.
+    ///
+    /// This will cause a panic if `num_cache_blocks + num_look_ahead_blocks < 3`.
     pub num_look_ahead_blocks: usize,
 
     /// The number of frames in a prefetch block.
     ///
     /// This should be left alone unless you know what you are doing.
-    pub block_size: usize,
+    pub block_frames: usize,
```
Comment (uklotzde) [inline core/src/read/mod.rs:89]: This is a breaking change for the API and requires to bump the major version to 2.

**Review by uklotzde — CHANGES_REQUESTED** [review]

**Comment by uklotzde** [comment]
> Hmm, there doesn't seem to be any way to revert to a previous version on crates.io. Oh well, I'll just go ahead and merge it.

There is no way back with version. Just increase the major version number on breaking changes.

**Comment by uklotzde** [comment]
All API changes should ideally be summarized in a CHANGELOG.md file

File: core/src/read/mod.rs (hunk 1 above)
Comment (BillyDM) [reply core/src/read/mod.rs:89]: Hmm, if we're going to bump it to version 2, then I might as well implement some other API changes I had planned as well. In theory there's no reason this couldn't also be used for network streaming, so I want to make things more generic to accommodate that. I'll go ahead and work on that then.

Comment (BillyDM) [reply core/src/read/mod.rs:89]: Though at first I think I'll at-least submit a PR that fixes the bug where the last remaining samples don't get written when a write stream is closed.

Comment (abique) [reply core/src/read/mod.rs:89]: From an outside perspective this is a good change, and better to do it the sooner rather than the later.

**Comment by BillyDM** [comment]
This PR is outdated now and a lot of these things have been fixed in separate PRs, so I think I'll close this. I might get around to revamping the API at some point, but I'd rather do that in a new PR with a fresh perspective.

I'll keep this branch around as reference, but I'll change the name to something like `dev_(outdated)`.

**Pull request closed without merging**